topdown/http.go

```diff
@@ -568,3 +622,349 @@ func parseTimeout(timeoutVal ast.Value) (time.Duration, error) {
+	respHeaders := getResponseHeaders(resp)
+	if respHeaders == nil {
+		return nil
+	}
+
+	var err error
+
+	pcv.date, err = getResponseHeaderDate(respHeaders)
+	if err != nil {
+		return err
+	}
+
+	pcv.cacheControl = parseCacheControlHeader(respHeaders)
+
+	pcv.expires, err = getResponseHeaderExpires(respHeaders)
+	if err != nil {
+		return err
+	}
+
+	pcv.etag = getResponseHeaderStrVal(respHeaders, "etag")
+
+	pcv.lastModified = getResponseHeaderStrVal(respHeaders, "last-modified")
+
+	return nil
+}
+
+func revalidateCachedResponse(req *http.Request, client *http.Client, resp *persistentCacheValue, forceJSONDecode bool) (ast.Value, int, bool, error) {
+	etag := resp.etag
+	lastModified := resp.lastModified
+
+	if etag == "" && lastModified == "" {
+		return nil, 0, false, nil
+	}
+
+	cloneReq := req.Clone(req.Context())
+
+	if etag != "" {
+		cloneReq.Header.Set("if-none-match", etag)
+	}
+
+	if lastModified != "" {
+		cloneReq.Header.Set("if-modified-since", lastModified)
+	}
+
+	// execute the http request
+	response, err := client.Do(cloneReq)
+	if err != nil {
+		return nil, 0, false, err
+	}
+	defer response.Body.Close()
+
+	switch response.StatusCode {
+	case http.StatusOK:
+		newResp, size, err := formatHTTPResponseToAST(response, forceJSONDecode)
+		return newResp, size, true, err
+
+	case http.StatusNotModified:
+		return resp.value, 0, false, nil
+	}
+
+	return nil, 0, false, nil
+}
+
+func canStore(resp ast.Value) bool {
+	respHeaders := getResponseHeaders(resp)
+	if respHeaders == nil {
+		return false
+	}
+
+	ccHeaders := parseCacheControlHeader(respHeaders)
+
+	// Check "no-store" cache directive
+	// The "no-store" response directive indicates that a cache MUST NOT
+	// store any part of either the immediate request or response.
+	if _, ok := ccHeaders["no-store"]; ok {
+		return false
+	}
+	return true
+}
+
+func isCachedResponseFresh(bctx BuiltinContext, resp *persistentCacheValue) bool {
+	if resp.date.IsZero() {
+		return false
+	}
+
+	currentTime := getCurrentTime(bctx)
+	if currentTime.IsZero() {
+		return false
+	}
+
+	currentAge := currentTime.Sub(resp.date)
+
+	// Check "max-age" cache directive.
+	// The "max-age" response directive indicates that the response is to be
+	// considered stale after its age is greater than the specified number
+	// of seconds.
+	if maxAge, ok := resp.cacheControl["max-age"]; ok {
+		maxAgeDur, err := time.ParseDuration(maxAge + "s")
+		if err != nil {
+			return false
+		}
+
+		if maxAgeDur > currentAge {
+			return true
+		}
+	} else {
+		// Check "Expires" header.
+		// Note: "max-age" if set, takes precedence over "Expires"
+		if resp.expires.Sub(resp.date) > currentAge {
+			return true
+		}
+	}
+	return false
+}
+
+func getCurrentTime(bctx BuiltinContext) time.Time {
+	var current time.Time
+
+	value, err := ast.JSON(bctx.Time.Value)
+	if err != nil {
+		return current
+	}
+
+	valueNum, ok := value.(json.Number)
+	if !ok {
+		return current
+	}
+
```

Comment:
Nit: I'd just cache the parsed duration on the persistentCacheValue instead of recomputing it each time.